internal/devbox/devbox.go

```diff
@@ -819,23 +804,32 @@ func (d *Devbox) ensureStateIsUpToDateAndComputeEnv(
 ) (map[string]string, error) {
 	defer debug.FunctionTimer().End()
 
-	// When ensureStateIsUpToDate is called with ensure=true, it always
-	// returns early if the lockfile is up to date. So we don't need to check here
-	if err := d.ensureStateIsUpToDate(ctx, ensure); isConnectionError(err) {
-		if !fileutil.Exists(d.nixPrintDevEnvCachePath()) {
-			ux.Ferrorf(
+	if envOpts.SkipRecompute {
+		upToDate, _ := d.lockfile.IsUpToDateAndInstalled(isFishShell())
+		if !upToDate {
+			if envOpts.Hooks.OnStaleStateWithSkipRecompute != nil {
+				envOpts.Hooks.OnStaleStateWithSkipRecompute()
+			}
+		}
```

Comment:
I was playing around with putting this outside the `envOpts.SkipRecompute` if statement.

It allows the hook to be more generic (i.e. `OnStaleState`). The downside is it can be a bit wasteful (calling `IsUpToDateAndInstalled` if we don't need to). `IsUpToDateAndInstalled` is fast (~1ms) but can be slower for large projects with remote plugins etc. So maybe we're all good as is. Just food for thought.

Reply:
Discussed in slack that this should be okay (for perf) to make generic into OnStaleState